PSA key identifiers rework

ronald-cron-arm opened the pull request:
## Description
This PR is the first part of the work related to #3265.

With openless PSA APIs, transient key identifier will be introduced for volatile keys thus the need to changes things around key identifiers in the libray. This PR doesn't introduce the transient key identifiers but pave the way to their introduction.

This PRs addresses two main things:
1) Always define `psa_key_id_t` as defined in PSA crypto spec. Before this PR, this is not the case when the MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER configuration flag is set.

2) Make the library compiles when the MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID (renaming of MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER) and test it with that configuration flag set
thus test the PSA APIs when the key identifiers are extended to contain a key owner identifier. 

## Status
Needs review and expected to need rework as well.

## Requires Backporting
No, PSA only work.

ronald-cron-arm commented:
The CI is as good as it can. Apart from Mbed OS Testing and Travis current issues, "merge TLS Testing" and "pr-merge" are falling because of the "ABI-API" testing as expected as this PR changes the definition of API parameter types.

paul-elliott-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -307,6 +328,10 @@ void create_existent( int lifetime_arg, int id_arg,
     size_t reexported_length;
     reopen_policy_t reopen_policy = reopen_policy_arg;
 
+    #if !defined(MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID)
```
Nit:Usual complaint about this belonging at the beginning of the function.

ronald-cron-arm replied:
Usual reply.

paul-elliott-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -437,17 +464,20 @@ void copy_across_lifetimes( int source_lifetime_arg, int source_id_arg,
     psa_algorithm_t expected_alg2 = expected_alg2_arg;
     uint8_t *export_buffer = NULL;
 
+    #if !defined(MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID)
```
Nit: Usual

ronald-cron-arm replied:
ditto

ronald-cron-arm commented on library/psa_crypto_storage.c:
```diff
@@ -62,22 +62,21 @@
+static psa_storage_uid_t psa_its_identifier_of_slot( PSA_KEY_ID_T key_id )
 {
-#if defined(MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER) && \
-    defined(PSA_CRYPTO_SECURE)
```
This is done to decouple the usage of extended identifiers from the PSA platform case to be able to run the unit tests with extended identifiers on the host machine. That's what I mean in the commit message by: "The changes in PSA code just make the enablement of extended key identifiers platform independent." I've tried to improve the commit message.

ronald-cron-arm commented on library/psa_crypto_storage.c:
```diff
@@ -62,22 +62,21 @@
      * the whole range 0..0xffffffff is available for non-key files. */
-    uint32_t unsigned_owner = (uint32_t) file_id.owner;
-    return( (uint64_t) unsigned_owner << 32 | file_id.key_id );
+    uint32_t unsigned_owner_id = (uint32_t) key_id.owner_id;
```
I've done it in a new commit.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -1312,19 +1314,24 @@ void static_checks( )
     size_t bits = bits_arg;
 
-    TEST_EQUAL( psa_get_key_id( &attributes ), 0 );
+    #if !defined(MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID)
```
I've tried to improve things.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -153,27 +153,27 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
- *                      0 to allow only key ids in the application range.
+ * \param key_id     The key identifier to check.
+ * \param vendor_ok  Nonzero to allow key identifiers in the vendor range.
+ *                   0 to allow only key dentifiers in the application range.
```
Fixed

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -1357,24 +1365,34 @@ void attributes_set_get( int id_arg, int lifetime_arg,
+                                                expected_id_arg );
     psa_key_lifetime_t expected_lifetime = expected_lifetime_arg;
 
+    #if !defined(MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID)
```
ditto, not there anymore.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_persistent_key.function:
```diff
@@ -139,16 +139,20 @@ exit:
     psa_key_type_t second_type = (psa_key_type_t) second_type_arg;
     psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
 
+    #if !defined(MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID)
```
not there anymore

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_persistent_key.function:
```diff
@@ -193,14 +197,18 @@ exit:
     psa_key_handle_t handle = 0;
     psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
 
+    #if !defined(MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID)
```
ditto

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_se_driver_hal.function:
```diff
@@ -761,12 +761,17 @@ exit:
     for( id = 0; id <= MAX_KEY_ID_FOR_TEST; id++ )
-        psa_destroy_persistent_key( id );
+    {
+        key_id = (PSA_KEY_ID_T)PSA_KEY_ID_INIT( 1, id );
```
The cast was not necessary. Removed.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_se_driver_hal.function:
```diff
@@ -1420,12 +1426,16 @@ void register_key_smoke_test( int lifetime_arg,
     psa_key_handle_t handle = 0;
     psa_status_t status;
 
+    #if !defined(MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID)
```
ditto

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_se_driver_hal_mocks.function:
```diff
@@ -90,12 +90,17 @@ static struct
     for( id = 0; id <= MAX_KEY_ID_FOR_TEST; id++ )
-        psa_destroy_persistent_key( id );
+    {
+        key_id = (PSA_KEY_ID_T)PSA_KEY_ID_INIT( 1, id );
```
Cast not necessary, removed.

ronald-cron-arm commented on include/mbedtls/config.h:
```diff
@@ -1260,20 +1260,20 @@
  */
 //#define MBEDTLS_ENTROPY_NV_SEED
 
-/* MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER
```
When the TF-M team updates to a new version of the present library it is easy for them to change such a define if necessary, changing it shouldn't be a problem for them.

ronald-cron-arm commented on include/psa/crypto.h:
```diff
@@ -57,6 +55,10 @@ extern "C" {
  * algorithms, key types, policies, etc. */
 #include "crypto_types.h"
 
+/* Include "crypto_platform.h" after "crypto_types.h" as it relies on
+ * `psa_key_id_t` defined by "crypto_types.h" */
+#include "crypto_platform.h"
```
I am not sure what do you mean here. What in crypto_types.h relies on crypto_platform.h? My understanding is that crypto_types.h contains the types defined by the PSA spec thus I don't really get why something in this file would be platform dependent. This was the case when `psa_key_id_t` was redefined on a platform basis but this is one of the goal of this PR to avoid that.

gilles-peskine-arm replied:
`psa_key_extended_id_t` (or whatever it ends up being called) has a definition that depends on the platform.

ronald-cron-arm replied:
This has been removed.

ronald-cron-arm commented on include/psa/crypto_platform.h:
```diff
@@ -46,57 +46,46 @@
 typedef int32_t psa_key_owner_id_t;
+#else
+/* Default */
+typedef uint32_t psa_key_owner_id_t;
```
yes it makes sense, done as suggested in commit "Fix PSA code and unit tests".

ronald-cron-arm commented on include/psa/crypto_platform.h:
```diff
@@ -46,57 +46,46 @@
+#endif /* !MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID */
 
-#endif /* !MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER */
+#define PSA_KEY_ID_EQUAL( id1, id2 ) \
```
changed in "Fix PSA code and unit tests" commit.

ronald-cron-arm commented on include/psa/crypto_platform.h:
```diff
@@ -46,57 +46,46 @@
+#else /* !MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID */
 
-#else /* !MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER */
+#define PSA_KEY_ID_T psa_key_id_t
```
I did it like that for two reasons:
1) That way the signature of the PSA APIs in crypto.h for openless APIS (will) look close to what is defined in the PSA specification. For example,
psa_status_t psa_sign_hash(PSA_KEY_ID_T key, ... ) instead of psa_status_t psa_sign_hash(psa_key_id_t key, ...) in the spec.

2) I didn't find a good name for a type that would be equal to `psa_key_id_t` if  MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER (or equivalent) is not defined and `psa_key_extended_id_t` (or equivalent) otherwise.

Thinking about this again, I think I'd prefer just having `psa_key_id_t` not defined as in the PSA spec (in case where key identifiers encode an owner identifier) and all the PSA openless API signatures defined as in the PSA spec, rather than the situation induced by the current status of this PR where `psa_key_id_t` is always as defined in the PSA spec but all the PSA openless API signatures are(will) not (again, in case where key identifiers encode an owner identifier). I'd prefer the first solution because there would just one deviation from the spec in the interface with expose. We already discussed about that and I think you were in favor of the second solution if I remember/understood correctly. Is it the case? Why?

gilles-peskine-arm replied:
1. With a macro, there's `PSA_KEY_ID_T` instead of `psa_key_id_t`. With a type, there's `psa_key_id_with_owner_t` instead of `psa_key_id_t`. Same difference.
2. With a type, you need to find a good name for the type. With a macro, you need to find a good name for the macro. Same difference.

I prefer to keep `psa_key_id_t` defined as in the PSA specification because I think it makes the code misleading otherwise. The reason `crypto.h` currently uses a redefined `psa_key_id_t` is that it was used as the source of the PSA specification document, and this reason no longer applies.

ronald-cron-arm replied:
Fixed

ronald-cron-arm commented on include/mbedtls/config.h:
```diff
@@ -1260,20 +1260,20 @@
 //#define MBEDTLS_ENTROPY_NV_SEED
 
-/* MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER
+/* MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID
```
What I am quite keen to change is to remove `file` in MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER and `psa_key_file_id_t`. I gave the main reason in the commit message of the commit "psa: Rename psa_key_file_id_t to psa_key_extended_id_t":
    
With PSA crypto v1.0.0, a volatile key identifier may contain a owner identifier but no file is associated to it. Thus rename the type psa_key_file_id_t to psa_key_extended_id_t to avoid a direct link with a file when a key identifier involves an owner
identifier. Rename also some related symbols.

If we stick to the goal of removing `file`, it could be MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER instead of `MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER`, and `psa_key_id_encodes_owner_t` instead of `psa_key_file_id_t`. What do you think?

gilles-peskine-arm replied:
I'm ok with `MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER`. To be revisited when we decide on a name for `struct {psa_key_owner_id_t; psa_key_id_t;}`.

ronald-cron-arm commented on include/psa/crypto_platform.h:
```diff
@@ -46,57 +46,46 @@
+#define PSA_KEY_ID_INIT( owner_id, owner_key_id ) { \
+            (psa_key_owner_id_t)( owner_id ), \
+            (psa_key_id_t)( owner_key_id ) }
+#define PSA_KEY_ID_GET_ID( key_id ) ( ( key_id ).owner_key_id )
```
I'll revisit this when we agree on the new names if any.

ronald-cron-arm replied:
Fixed

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -487,7 +484,20 @@ void copy_across_lifetimes( int source_lifetime_arg, int source_id_arg,
+        #if defined(MBEDTLS_PSA_CRYPTO_KEY_EXTENDED_ID)
+        TEST_EQUAL( PSA_KEY_ID_GET_OWNER_ID( target_id ), 1 );
+        #endif
+        TEST_EQUAL( PSA_KEY_ID_GET_ID( target_id ) , 0 );
```
Because we need to construct the (1,0) identifier and then compare it with `target_id`. I can do it if you prefer, not done yet.

gilles-peskine-arm replied:
I was wondering because in other places, you do use `PSA_KEY_ID_EQUAL`, and I couldn't find a reason for doing it differently.

I see advantages to both approaches. `PSA_KEY_ID_EQUAL` respects the abstraction, so it's more robust. Testing the two parts separately will give a better error message once we finally make `TEST_EQUAL` print the values.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_se_driver_hal.function:
```diff
@@ -761,12 +761,17 @@ exit:
     psa_key_location_t location;
+
     /* The tests may have potentially created key ids from 1 to
      * MAX_KEY_ID_FOR_TEST. In addition, run the destroy function on key id
      * 0, which file-based storage uses as a temporary file. */
```
Just started to look at it. Not addressed yet.

ronald-cron-arm replied:
Addressed in two additional commits.

ronald-cron-arm commented:
I've rebased on top of development and addressed most of the issues, not all though. The previous version of the PR can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/key-extended-id-1).

ronald-cron-arm commented:
The CI testing is back to "as good as it can". Apart from Mbed OS Testing current issues, "merge TLS Testing" and "pr-merge" are falling because of the "ABI-API" testing as expected as this PR changes the definition of API parameter types.

@paul-elliott-arm I've addressed all your comments.

@gilles-peskine-arm I've addressed most of your comments. For the remaining ones (renaming and include order), I've tried to clarify the intent or asked questions in trying to find a way forward. Please have a look.

I am putting back the label "needs: review" as I am now waiting for replies to my clarifications and questions.

gilles-peskine-arm commented on include/psa/crypto_platform.h:
```diff
@@ -44,57 +44,91 @@
+ * \param owner_id Identifier of the key owner.
+ * \param key_id   Identifier of the key.
+ */
+static inline psa_key_extended_id_t psa_key_id_init(
```
I'm not sure I like this name because `psa_key_id_init(owner_id, key_id)` is only loosely related to `PSA_KEY_ID_INIT` (which takes no argument). But I don't really have a better name to propose either. `psa_key_id_make`?

We don't have a naming convention for functions like this because we don't yet have functions like this, so this is setting a precedent.

ronald-cron-arm replied:
I've changed to *_make.

ronald-cron-arm commented:
I've addressed all comments and the CI is back to as good as it can. Please review again. Previous reviewed version of the PR can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/key-extended-id-2).

gilles-peskine-arm commented on include/psa/crypto_platform.h:
```diff
@@ -44,57 +44,33 @@
 /* PSA requires several types which C99 provides in stdint.h. */
 #include <stdint.h>
 
+#if ( defined(__ARMCC_VERSION) || defined(_MSC_VER) ) && \
+    !defined(inline) && !defined(__cplusplus)
+#define inline __inline
+#endif
+
```
`crypto_struct.h` already contained `static line` functions and neither our CI nor any user have complained. Maybe it only got away with this accidentally because nobody has included it in a configuration that didn't also pull in `mbedtls/cipher.h` or `mbedtls/pk.h`. But in principle, this should no longer be necessary since we require C99. So please don't add it.

ronald-cron-arm replied:
I added this due to a build falling in the CI. As things have changes quite significantly let's remove it and see if the CI still complains.

ronald-cron-arm replied:
Some Windows builds fail without this thus I've added it back.

gilles-peskine-arm commented on include/psa/crypto_types.h:
```diff
@@ -226,15 +233,84 @@ typedef uint32_t psa_key_location_t;
+#if !defined(MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER)
+typedef psa_key_id_t mbedtls_svc_key_id_t;
+
+#define MBEDTLS_SVC_KEY_ID_INIT ( (psa_key_id_t)0 )
```
Not very important since the distinction between `crypto_types.h` and `crypto_values.h` is purely cosmetic, but in principle, macros and functions to analyze or construct values should be in `crypto_values.h`, not `crypto_types.h`.

ronald-cron-arm replied:
Better to put away things as they should, thus kept in crypto_types.h just the type definitions and moved the rest to crypto_values.h.

gilles-peskine-arm reviewed (changes requested):
I'm happy overall. I only have a few localized comments.

ronald-cron-arm commented on scripts/config.py:
```diff
@@ -248,6 +247,7 @@ def full_adapter(name, active, section):
     'MBEDTLS_PLATFORM_TIME_ALT', # requires a clock and HAVE_TIME
     'MBEDTLS_PSA_CRYPTO_SE_C', # requires a filesystem and PSA_CRYPTO_STORAGE_C
     'MBEDTLS_PSA_CRYPTO_STORAGE_C', # requires a filesystem
+    'MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER', # requires a multi-client service
```
This is done in the "Set MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER in full configuration" commit. This prevents it to be added to `baremetal` config.

gilles-peskine-arm replied:
Oh right, this is the baremetal list. But why doesn't it work in baremetal? The given reason “requires a multi-client service” no longer applies.

ronald-cron-arm replied:
That's right the reason no longer applies. I've just tried the baremetal config with MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER enabled and some unit tests are falling (hadn't tried before). Having a look to understand why some unit tests are falling with this configuration.

ronald-cron-arm replied:
In fact, the unit tests are failing with the baremetal configuration and in the CI we do only build tests of the baremetal configuration. Finally, adding MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER  to the baremetal configuration.

ronald-cron-arm commented on include/psa/crypto_types.h:
```diff
@@ -37,6 +37,11 @@
 
 #include <stdint.h>
 
+#if ( defined(__ARMCC_VERSION) || defined(_MSC_VER) ) && \
+    !defined(inline) && !defined(__cplusplus)
+#define inline __inline
+#endif
+
```
Removed.

ronald-cron-arm commented on include/mbedtls/config.h:
```diff
@@ -1258,20 +1258,20 @@
- * multi-client service. When you activate this option, you must provide
- * an implementation of the type psa_key_owner_id_t and a translation
- * from psa_key_file_id_t to file name in all the storage backends that
+ * multi-client service. When you activate this option, you must provide an
+ * implementation of the type mbedtls_key_owner_id_t and a translation from
+ * mbedtls_svc_key_id_t to file name in all the storage backends that you
  * you wish to support.
```
Tried to fix things in a new commit.

ronald-cron-arm commented:
@gilles-peskine-arm and @paul-elliott-arm, CI is back to as good as it can and I've addressed all comments. Please have another look.

paul-elliott-arm reviewed (approved):
Looks good to me!
